File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/adapter/SectionsAdapter.java
package home.oleg.placesnearme.feature_map.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import home.oleg.placenearme.models.Section;
import home.oleg.placesnearme.feature_map.R;
import home.oleg.placesnearme.feature_map.viewdata.SectionViewData;

/**
 * Created by Oleg Sheliakin on 25.10.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public class SectionsAdapter extends RecyclerView.Adapter<SectionsAdapter.ViewHolder> implements ItemSelectListener {

    private final List<CheckedItem<SectionViewData>> checkedItems;
    private final SectionSelectListener sectionSelectListener;

    public SectionsAdapter(List<CheckedItem<SectionViewData>> checkedItems, SectionSelectListener sectionSelectListener) {
        this.checkedItems = checkedItems;
        this.sectionSelectListener = sectionSelectListener;
    }

    public int getSelectedItemPosition() {
        int selectedItemPosition = -1;
        for (int i = 0; i < checkedItems.size(); i++) {
            if (checkedItems.get(i).isChecked()) {
                selectedItemPosition = i;
            }
        }
        return selectedItemPosition;
    }

    public Section getSelectedItem() {
        int selectedItemPosition = getSelectedItemPosition();
        if (selectedItemPosition != -1) {
            return checkedItems.get(selectedItemPosition).getData().getSection();
        }
        return null;
    }

    public interface SectionSelectListener {
        void sectionSelected(Section section);
    }

    @NonNull
    @Override
    public SectionsAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_section, viewGroup, false);
        return new SectionsAdapter.ViewHolder(view, this);
    }

    @Override
    public void onBindViewHolder(@NonNull SectionsAdapter.ViewHolder viewHolder, int i) {
        viewHolder.bind(checkedItems.get(i));
    }

    @Override
    public int getItemCount() {
        return checkedItems.size();
    }

    @Override
    public void itemSelected(int position) {
        setSelected(position);
        sectionSelectListener.sectionSelected(checkedItems.get(position).getData().getSection());
    }

    public void setSelected(int position) {
        if (position == -1) {
            return;
        }

        for (CheckedItem<SectionViewData> checkedItem : checkedItems) {
            checkedItem.setChecked(false);
        }
        checkedItems.get(position).setChecked(true);
        notifyDataSetChanged();
    }

    static final class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView textView;

        private ViewHolder(@NonNull View itemView, ItemSelectListener itemSelectListener) {
            super(itemView);
            textView = itemView.findViewById(R.id.tvSection);
            textView.setOnClickListener(view -> itemSelectListener.itemSelected(getAdapterPosition()));
        }

        private void bind(CheckedItem<SectionViewData> checkedItem) {
            SectionViewData sectionViewData = checkedItem.getData();
            textView.setSelected(checkedItem.isChecked());
            textView.setText(sectionViewData.getSectionNameRes());
            textView.setCompoundDrawablesWithIntrinsicBounds(0, sectionViewData.getSectionDrawableRes(), 0, 0);
        }
    }
}

File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/presentation/ui/adapter/SectionsAdapter.kt
package home.oleg.placesnearme.feature_map.presentation.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import androidx.recyclerview.widget.RecyclerView
import home.oleg.placenearme.models.Section
import home.oleg.placesnearme.feature_map.R
import home.oleg.placesnearme.feature_map.presentation.sections.SectionViewData

/**
 * Created by Oleg Sheliakin on 25.10.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
class SectionsAdapter(
        private val checkedItems: List<CheckedItem<SectionViewData>>,
        private val sectionSelectListener: SectionSelectListener)
    : RecyclerView.Adapter<SectionsAdapter.ViewHolder>(), ItemSelectListener {

    val selectedItemPosition: Int
        get() {
            var selectedItemPosition = -1
            for (i in checkedItems.indices) {
                if (checkedItems[i].isChecked) {
                    selectedItemPosition = i
                }
            }
            return selectedItemPosition
        }

    val selectedItem: Section?
        get() {
            val selectedItemPosition = selectedItemPosition
            return if (selectedItemPosition != -1) {
                checkedItems[selectedItemPosition].data.section
            } else null
        }

    interface SectionSelectListener {
        fun sectionSelected(section: Section)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.item_section, viewGroup, false)
        return SectionsAdapter.ViewHolder(view, this)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        viewHolder.bind(checkedItems[i])
    }

    override fun getItemCount(): Int {
        return checkedItems.size
    }

    override fun getItemId(position: Int): Long = position.toLong()

    override fun itemSelected(position: Int) {
        setSelected(position)
        sectionSelectListener.sectionSelected(checkedItems[position].data.section)
    }

    fun setSelected(position: Int) {
        if (position == -1) {
            return
        }

        for (checkedItem in checkedItems) {
            checkedItem.isChecked = false
        }
        checkedItems[position].isChecked = true
        notifyDataSetChanged()
    }

    class ViewHolder constructor(itemView: View, itemSelectListener: ItemSelectListener) : RecyclerView.ViewHolder(itemView) {

        private val textView: TextView = itemView.findViewById(R.id.tvSection)

        init {
            textView.setOnClickListener { itemSelectListener.itemSelected(adapterPosition) }
        }

        fun bind(checkedItem: CheckedItem<SectionViewData>) {
            val (_, sectionNameRes, sectionDrawableRes) = checkedItem.data
            textView.isSelected = checkedItem.isChecked
            textView.setText(sectionNameRes)
            textView.setCompoundDrawablesWithIntrinsicBounds(0, sectionDrawableRes, 0, 0)
        }
    }
}
